### PR TITLE
[amp-story-player] Implement `go` method

### PR DESCRIPTION
**Pull request by bettyashley**
Implement internal callback to navigate stories given a number.

/cc @Enriqe @newmuis 

Partial for: #28416

### CODE REVIEW

File: test/unit/test-amp-story-player.js [hunk 1]
```diff
@@ -370,5 +370,49 @@ describes.realWin('AmpStoryPlayer', {amp: false}, (env) => {
...
+      await player.load();
+
+      const stories = toArray(playerEl.querySelectorAll('a'));
+
+      player.go(0);
+      expect(stories[0][IFRAME_IDX]).to.eql(0);
+      player.go(1);
+      expect(stories[0][IFRAME_IDX]).to.eql(1);
+      player.go(-1);
+      expect(stories[0][IFRAME_IDX]).to.eql(0);
```
Comment (bettyashley) [inline test/unit/test-amp-story-player.js:389]: This test does not currently pass, I am unsure how to check the story's iframe index property.

Comment (Enriqe) [reply test/unit/test-amp-story-player.js:389]: So what `IFRAME_IDX` tells you is the iframe index that corresponds to a given story (if any), but this doesn't really tell you which of those is the currently active iframe. 

I guess a more appropriate check here would be to check the iframe position, like we do in other tests:

https://github.com/ampproject/amphtml/blob/33e0c090956904a9adcf1d877b4c7727a6352543/test/unit/test-amp-story-player.js#L219

Comment (Enriqe) [reply test/unit/test-amp-story-player.js:389]: When iframe position is `0`, it is the currently displayed story, when it is either `1` or `-1` it means the iframe is at either side of the story.

File: src/amp-story-player/amp-story-player-impl.js [hunk 2]
```diff
@@ -486,6 +487,36 @@ export class AmpStoryPlayer {
...
+    if (this.currentIdx_ + storyDelta < 0) {
+      throw new Error('Out of Story range.')
+    }
+    
+    if (storyDelta === 0) {
+      return;
+    }
+    if (storyDelta > 0) {
+      for (let i = 0; i < storyDelta; i++) {
+        this.next_();
```
Comment (newmuis) [inline src/amp-story-player/amp-story-player-impl.js:508]: @Enriqe will this cause multiple allocations/deallocations of iframes?

File: test/unit/test-amp-story-player.js (hunk 1 above)
Comment (bettyashley) [reply test/unit/test-amp-story-player.js]: I just changed the test, it still currently fails. But let know if I'm on the right track please.

File: src/amp-story-player/amp-story-player-impl.js [hunk 3]
```diff
@@ -486,6 +487,36 @@ export class AmpStoryPlayer {
     }
   }
 
+  /**
+   * Navigates stories given a number.
+   * @param {number} storyDelta
+   * @private
+   */
+  go(storyDelta) {
+    if (this.currentIdx_ + storyDelta > this.stories_.length) {
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:496]:
```suggestion
    if (this.currentIdx_ + storyDelta >= this.stories_.length) {
```

File: src/amp-story-player/amp-story-player-impl.js [hunk 4]
```diff
@@ -486,6 +487,36 @@ export class AmpStoryPlayer {
...
+      throw new Error('Out of Story range.')
+    }
+    if (this.currentIdx_ + storyDelta < 0) {
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:499]: We can join this check with the `if` above.

File: src/amp-story-player/amp-story-player-impl.js [hunk 5]
```diff
@@ -486,6 +487,36 @@ export class AmpStoryPlayer {
...
+      throw new Error('Out of Story range.')
+    }
+    
+    if (storyDelta === 0) {
+      return;
+    }
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:505]:
```suggestion
    }
    
```

File: src/amp-story-player/amp-story-player-impl.js (hunk 2 above)
Comment (Enriqe) [reply src/amp-story-player/amp-story-player-impl.js:508]: Yeah, especially if the given delta is a story further in the array that has many stories in between without an iframe. I think we should be able to optimize this by using the `show()` method instead. @bettyashley

File: test/unit/test-amp-story-player.js [hunk 6]
```diff
@@ -370,5 +370,45 @@ describes.realWin('AmpStoryPlayer', {amp: false}, (env) => {
...
+      appendStoriesToPlayer(playerEl, 5);
+
+      const player = new AmpStoryPlayer(win, playerEl);
+
+      await player.load();
+
+      const iframes = playerEl.shadowRoot.querySelectorAll('iframe');
+
+      player.go(0);
+      expect(iframes[0].getAttribute('i-amphtml-iframe-position')).to.eql('0');
```
Comment (Enriqe) [inline test/unit/test-amp-story-player.js:385]: Did you try wrapping these inside a `win.requestAnimationFrame(() => {})` like the other tests to see if that helps?

https://github.com/ampproject/amphtml/blob/33e0c090956904a9adcf1d877b4c7727a6352543/test/unit/test-amp-story-player.js#L216-L220

**Comment by bettyashley** [comment]
A couple of errors with the updated code:
- the updated go method passes the test but does not work properly in a demo
- wrapping the test worked as you said @Enriqe 
- when I run my three tests, only the first one runs at all (and passes), I have to run the other two separately and they pass, not sure if this is an indication of something being wrong

File: src/amp-story-player/amp-story-player-impl.js [hunk 7]
```diff
@@ -486,6 +487,40 @@ export class AmpStoryPlayer {
     }
   }
 
+  /**
+   * Navigates stories given a number.
+   * @param {number} storyDelta
+   * @private
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:493]:
```suggestion
```

File: src/amp-story-player/amp-story-player-impl.js [hunk 8]
```diff
@@ -486,6 +487,40 @@ export class AmpStoryPlayer {
...
+  go(storyDelta) {
+    if (this.currentIdx_ + storyDelta >= this.stories_.length || this.currentIdx_ + storyDelta < 0) {
+      throw new Error('Out of Story range.')
+    }
+    
+    if (storyDelta === 0) {
+      return;
+    }
+
+    const newIndex = this.stories_[this.currentIdx_ + storyDelta];
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:504]: `this.stories_` is an array containing anchor elements that represent a story (`<a href="story.html">`), so `newIndex` here is not actually an index but an element. So the line below will not be doing what you want it to do.

File: src/amp-story-player/amp-story-player-impl.js [hunk 9]
```diff
@@ -486,6 +487,39 @@ export class AmpStoryPlayer {
...
+
+    const currentStory = this.stories_[newIndex];
+    this.updateCurrentIframe_(currentStory[IFRAME_IDX]);
+
+    const nextStoryIdx = newIndex + 1;
+    if (
+      nextStoryIdx < this.stories_.length &&
+      this.stories_[nextStoryIdx][IFRAME_IDX] === undefined
+    ) {
+      this.allocateIframeForStory_(nextStoryIdx);
```
Comment (Enriqe) [inline src/amp-story-player/amp-story-player-impl.js:519]: I'm not sure this will work as it is. The way this function is implemented now, it will deallocate an iframe from the story furthest away and rotate the iframe pool with the assumption that we are moving in a linear fashion and not skipping stories. I think if we test all the edge cases we will find something that breaks.

IMO we have two options:

1) use the show method here so that it can take care of this
2) update the `allocateIframeForStory_` method to take into account this edge case.

Comment (bettyashley) [reply src/amp-story-player/amp-story-player-impl.js]: Updated to use the show method!

File: test/unit/test-amp-story-player.js [hunk 10]
```diff
@@ -370,5 +370,47 @@ describes.realWin('AmpStoryPlayer', {amp: false}, (env) => {
...
+
+      win.requestAnimationFrame(() => {
+        const iframes = playerEl.shadowRoot.querySelectorAll('iframe');
+        
+        player.go(0);
+        expect(iframes[0].getAttribute('i-amphtml-iframe-position')).to.eql('0');
+        player.go(1);
+        expect(iframes[1].getAttribute('i-amphtml-iframe-position')).to.eql('0');
+        player.go(-1);
+        expect(iframes[0].getAttribute('i-amphtml-iframe-position')).to.eql('0');
```
Comment (gmajoulet) [inline test/unit/test-amp-story-player.js:390]: Please split these in three tests, so when things fail it's easier to know if the error is when navigating backward or forward. It's good unit testing practice to only assert one thing per test.

**Comment by gmajoulet** [comment]
@Enriqe are we tracking the work for a second `pageDelta` parameter somewhere?

**Comment by Enriqe** [comment]
Yep, just filed #29220

File: test/unit/test-amp-story-player.js [hunk 11]
```diff
@@ -370,5 +370,78 @@ describes.realWin('AmpStoryPlayer', {amp: false}, (env) => {
...
+      const player = new AmpStoryPlayer(win, playerEl);
+
+      await player.load();
+
+      win.requestAnimationFrame(() => {
+        const iframes = playerEl.shadowRoot.querySelectorAll('iframe');
+        player.go(2);
+        expect(iframes[1].getAttribute('i-amphtml-iframe-position')).to.eql(-1);
+        expect(iframes[2].getAttribute('i-amphtml-iframe-position')).to.eql(0);
+        expect(iframes[3].getAttribute('i-amphtml-iframe-position')).to.eql(1);
```
Comment (Enriqe) [inline test/unit/test-amp-story-player.js:387]: I don't think there's a 4th iframe. There's maximum 3.

File: test/unit/test-amp-story-player.js [hunk 12]
```diff
@@ -370,5 +370,78 @@ describes.realWin('AmpStoryPlayer', {amp: false}, (env) => {
...
+
+    it('navigate forward given a positive number in range', async () => {
+      const playerEl = win.document.createElement('amp-story-player');
+      appendStoriesToPlayer(playerEl, 5);
+
+      const player = new AmpStoryPlayer(win, playerEl);
+
+      await player.load();
+
+      win.requestAnimationFrame(() => {
```
Comment (Enriqe) [inline test/unit/test-amp-story-player.js:382]: I think the code inside this `requestAnimationFrame()` is not running correctly in the tests, I sent out #29273 to fix this.

File: test/unit/test-amp-story-player.js (hunk 11 above)
Comment (newmuis) [reply test/unit/test-amp-story-player.js:387]: Did the test still pass?

Comment (Enriqe) [reply test/unit/test-amp-story-player.js:387]: (See comment below) I think there's an issue with the tests and this code was not actually running

Comment (gmajoulet) [reply test/unit/test-amp-story-player.js:387]: We could either return a Promise that resolves after these `expect` statements, or get a `done` parameter from the `it()` second callback parameter, that we call after the `expect` statements.
`it('navigate forward given a positive number in range', async (done) => {`
